microcosm_pubsub/sentry.py

```diff
@@ -33,41 +53,80 @@ def before_send(event, hint):
                 if var_name in ["self", "cls"]:
                     continue
                 elif isinstance(value, str):
-                    if value.startswith("<function "):
+                    if value.startswith("<") and value.endswith(">"):
                         continue
                     elif var_name.endswith("_id"):
                         continue
+                    elif value.startswith("http"):
+                        continue
+                elif isinstance(value, dict):
+                    for key in value.keys():
+                        if key.endswith("_id"):
+                            continue
+                        value[key] = "<redacted>"
+                    continue
                 frame["vars"][var_name] = "<redacted>"
     return event
 
 
 @defaults(
     enabled=False,
     dsn=None,
+    custom_tags_mapping=None,
+    custom_before_send_func=None,
+    custom_user_id=None,
 )
 @logger
-def configure_sentry(graph: ObjectGraph):
-    enabled, dsn, sentry = False, None, None
+def configure_sentry(graph: ObjectGraph) -> SentryConfig:
+    enabled, dsn, sentry,  = False, None, None
+    tag_mapping, custom_before_send, custom_user_id_key = DEFAULT_TAG_MAPPING, None, DEFAULT_USER_ID_KEY
+
     if graph.config.sentry_logging.enabled and graph.config.sentry_logging.dsn and sentry_sdk:
         enabled = graph.config.sentry_logging.enabled
         dsn = graph.config.sentry_logging.dsn
         if graph.metadata.testing:
             sentry = MagicMock()
         else:
+            if graph.config.sentry_logging.custom_before_send_func:
+                custom_before_send = import_custom_before_send(graph)
+
             try:
                 sentry = sentry_sdk.init(
                     graph.config.sentry_logging.dsn,
                     server_name=f"{graph.metadata.name}_daemon",
                     release=graph.config.build_info_convention.sha1,
                     environment=environ.get("MICROCOSM_ENVIRONMENT") or "undefined",
-                    before_send=before_send,
+                    before_send=custom_before_send or before_send,
                 )
             except BadDsn:
                 enabled, dsn, sentry = False, None, None
                 configure_sentry.logger.error("Bad DSN value set")
 
+    if graph.config.sentry_logging.custom_tags_mapping:
+        tag_mapping.update(graph.config.sentry_logging.custom_tags)
+    if graph.config.sentry_logging.custom_user_id:
+        custom_user_id_key = graph.config.sentry_logging.custom_user_id
+
     return SentryConfig(
         dsn=dsn,
         enabled=enabled,
-        client=sentry
+        client=sentry,
+        tag_mapping=tag_mapping,
+        user_id_key=custom_user_id_key
     )
+
+
+def import_custom_before_send(graph: ObjectGraph) -> Optional[Callable[[Any, Any], Any]]:
+    func_path = graph.config.sentry_logging.custom_before_send_func
```

Comment:
can the before send func be a binding instead? this seems like it's a good usecase for our DI framework

Reply:
Ah good point! Will investigate